crates/datasources/src/clickhouse/mod.rs

```diff
@@ -117,32 +150,18 @@ impl ClickhouseAccessState {
                     SqlType::Float64 => DataType::Float64,
                     SqlType::String | SqlType::FixedString(_) => DataType::Utf8,
                     // Clickhouse has both a 'Date' type (2 bytes) and a
-                    // 'Date32' type (4 bytes). I think they're both represented
-                    // with this one variant.
+                    // 'Date32' type (4 bytes). The library doesn't support
+                    // 'Date32' type yet.
+                    // TODO: Maybe upstream support for Date32?
                     SqlType::Date => DataType::Date32,
                     SqlType::DateTime(DateTimeType::DateTime32) => {
-                        DataType::Timestamp(TimeUnit::Second, None)
+                        DataType::Timestamp(TimeUnit::Nanosecond, None)
                     }
-                    SqlType::DateTime(DateTimeType::DateTime64(precision, tz)) => {
-                        // Precision represents the tick size, computed as
-                        // 10^precision seconds.
-                        //
-                        // <https://clickhouse.com/docs/en/sql-reference/data-types/datetime64>
-                        let unit = match *precision {
-                            0 => TimeUnit::Second,
-                            3 => TimeUnit::Millisecond,
-                            6 => TimeUnit::Microsecond,
-                            9 => TimeUnit::Nanosecond,
-                            other => {
-                                return Err(ClickhouseError::String(format!(
-                                    "unsupported time precision: {other}"
-                                )))
-                            }
-                        };
-
-                        let tz: Arc<str> = tz.to_string().into_boxed_str().into();
-
-                        DataType::Timestamp(unit, Some(tz))
+                    SqlType::DateTime(DateTimeType::DateTime64(_precision, tz)) => {
+                        // We store all timestamps in nanos, so the precision
+                        // here doesn't matter. clickhouse_rs crate handles this
+                        // for us.
+                        DataType::Timestamp(TimeUnit::Nanosecond, Some(tz.to_string().into()))
```

Comment:
it matters, I think, more what _they_ store timestamps as; and it doesn't look like we're giving (or getting) information from `clickhouse_rs` about what they store as.

Reply:
Nanos is the most precise clickhouse can go. It’s essentially a design choice that we are making here. It’s easier if the whole codebase uses nanos instead of anything else. We return accurate values but yeah, we don’t exactly translate clickhouse. But this doesn’t seem like an issue (to me). Wondering… are there any real disadvantages to this?

Reply:
What gets easier if our entire codebase returns nanos instead of anything else?

I think we should maintain whatever precision the data source natively has and not try to cast numbers around to make things more consistent..

It seems like we could get ourselves into unexpected overflows or have users get unexpected results if their timestamp resolution changes, mostly for the cost of a few arms on a match statement, but I might be missing something.

Reply:
> What gets easier if our entire codebase returns nanos instead of anything else?

There’s a lot of stuff (IIRC) that DF has issues with (when it comes to types). Even this change is something I made when I encountered an error. But that’s a good point that we might get a weird overflow someday so I’ll create a different issue for now.

This might not be very convincing but we do have: https://github.com/GlareDB/glaredb/blob/bf86b29349fd319754ef4b7dca76162fc232fd5c/crates/datasources/src/common/util.rs#L141-L163

This existed for a good reason which in hindsight I should have noted down…

Reply:
Opened https://github.com/GlareDB/glaredb/issues/2369